tools/generator/dfg/rp/rp_device_tree.py

```diff
@@ -0,0 +1,283 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2013-2016, Niklas Hauser
+# Copyright (c)      2016, Fabian Greif
+# Copyright (c)      2020, Hannes Ellinger
+# All rights reserved.
+
+import math
+import logging
+import re
+import os
+
+from ..device_tree import DeviceTree
+from ..input.xml import XMLReader
+
+from .rp_identifier import RPIdentifier
+
+LOGGER = logging.getLogger('dfg.rp.reader')
+
+class RPDeviceTree:
+    """ RPDeviceTree
+    This RP specific part description file reader knows the structure and
+    translates the data into a platform independent format.
+    """
+
+    @staticmethod
+    def _properties_from_file(svd_filename):
+        device_file = XMLReader(svd_filename)
+        p = {}
+
+        partname = svd_filename.split('/')[-1]
+        partname = partname.split('.')[0]
+
+        did = RPIdentifier.from_string(partname.lower())
+        p['id'] = did
+
+        LOGGER.info("Parsing '%s'", did.string)
+
+        # information about the core and architecture
+        core = device_file.query("//device/cpu/name")[0].text.lower().replace("cm", "cortex-m")
+        if device_file.query("//device/cpu/fpuPresent")[0].text == '1':
+            core += "f"
+        p["core"] = core
+
+        # @todo
+        memories = [
+            # FLASH(rx) : ORIGIN = 0x10000000, LENGTH = 2048k
+            # RAM(rwx) : ORIGIN =  0x20000000, LENGTH = 256k
+            # SCRATCH_X(rwx) : ORIGIN = 0x20040000, LENGTH = 4k
+            # SCRATCH_Y(rwx) : ORIGIN = 0x20041000, LENGTH = 4k
+            {
+                "name": "ram",
+                "access": "rwx",
+                "size": str(int("0x40000",16)),
+                "start": "0x20000000"
+            },
+            {
+                "name": "scratch_x",
+                "access": "rwx",
+                "size": str(int("0x1000",16)),
+                "start": "0x20040000"
+            },
+            {
+                "name": "scratch_y",
+                "access": "rwx",
+                "size": str(int("0x1000",16)),
+                "start": "0x20041000"
+            },
```

Comment:
I think we should use the partition SRAM0 - SRAM5 as in the datasheet.

![](https://user-images.githubusercontent.com/163066/148319330-d901fc17-2dea-4eba-becf-ebb2c9108afd.png)

Reply:
for now using interleaved intersection mapping of SRAM0-SRAM4 as at pico-sdk (see page 148 for details). Usage separate banks more complex task and can be achieved by user with providing custom linker script.

Reply:
Oh wow, that's an interesting design, rerouting every 32-bit word to another bank!

I feel like we should rename `scratch_x/y` to sram4/5 though, since if I understand correctly the pico-sdk are used similarly to `modm_fastcode`/`modm_fastdata`, but there's no explicit reason to call it scratch instead of sram?

```
/*! \brief Section attribute macro for placement in the SRAM bank 4 (known as "scratch X")
 *  \ingroup pico_platform
 *
 * Scratch X is commonly used for critical data and functions accessed only by one core (when only
 * one core is accessing the RAM bank, there is no opportunity for stalls)
 *
 * For example a `uint32_t` variable placed in "scratch X"
 *
 *     uint32_t __scratch_x("my_group_name") foo = 23;
 *
 * The section attribute is `.scratch_x.<group>`
 *
 * \param group a string suffix to use in the section name to distinguish groups that can be linker
 *              garbage-collected independently
 */
#define __scratch_x(group) __attribute__((section(".scratch_x." group)))
```

Reply:
I think sram4/5 is not good names, if RPI create processor with more ram probably indexes changed. modm_fastcode / modm_fastdata placed to interleaved ram. scratch_x / scratch_y can be used only manually by user with care about usage on one core.